lib/python/Jetson/GPIO/gpio_event.py

```diff
@@ -48,195 +66,287 @@
 # epoll blocking wait object
 _epoll_fd_blocking = None
 
-# dictionary of GPIO class objects. key = gpio number,
+# 2-layered dictionary of GPIO class objects. 
+# layer 1 key = chip name, layer 2 key = channel (pin number by mode)
 # value = GPIO class object
 _gpio_event_list = {}
 
-# variable to keep track of thread state
-_thread_running = False
-
-# string representations for edges to write to sysfs
-_edge_str = ["none", "rising", "falling", "both"]
+# Dictionary for thread to lookup to check its supposed state
+# Key: thread id, Value: true if it is supposed to be running state, otherwise false
+_thread_running_dict = {}
 
 # lock object for thread
 _mutex = thread.allocate_lock()
 
-
 class _Gpios:
-    def __init__(self, gpio_name, edge=None, bouncetime=None):
-        self.edge = edge
-        self.value_fd = open(ROOT + "/" + gpio_name + "/value", 'r')
+    # @value_fd the file descriptor for the chip line
+    # @initial_thread true if the thread just start up (within the first loop)
+    # @thread_added the number of threads being added to monitor this object/gpio 
+    # @thread_id the id of the thread being created to detect event
+    # @bouncetime the time interval for debouncing
+    # @callbacks a list of callback functions to be executed when an edge event happened
+    # @lastcall the timestamp for counting debounce
+    # @event_occurred true if an edge event occured
+    def __init__(self, line_fd, bouncetime=None):
+        self.value_fd = line_fd 
         self.initial_thread = True
-        self.initial_wait = True
         self.thread_added = False
+        self.thread_id = 0
         self.bouncetime = bouncetime
         self.callbacks = []
         self.lastcall = 0
         self.event_occurred = False
 
     def __del__(self):
-        self.value_fd.close()
         del self.callbacks
 
-
-def add_edge_detect(gpio, gpio_name, edge, bouncetime):
+# @brief adding an edge detecting event 
+#   The detection event runs in an thread that enables non-blocking I/O multiplexing approach.
+#   However, one pin on a chip (channel) can only allow one edge detection event, the new added
+#   event will be removed if there's an existing event.
+# @param[in] chip_fd: file descriptor 
+# @param[in] chip_name: the GPIO chip name/instance
+# @param[in] channel: the file descriptor for the chip line
+# @param[in] request: gpioevent_request struct that describes gpio event monitoring
+# @param[in] bouncetime: the time interval for debouncing
+# @param[in] poll_time: the max time to wait for an edge event
+# @param[out] success on 0, otherwise return 2 if something fatal happened
+def add_edge_detect(chip_fd, chip_name, channel, request, bouncetime, poll_time):
```

Comment:
We should have a default poll_time for compatibility purpose. and print a warning if they use this without a poll_timeout so people who is transiting from the old code can update